[Rae2480] iP

Rae2480 opened the pull request:
### Viktory

> "It's not enough to give people what they need to survive. You gotta give them what they need to live." - Ekko (Arcane)

Are you...
- [ ]  familiar with the hit netflix series [Arcane](https://en.wikipedia.org/wiki/Arcane_(TV_series))?
- [ ] Riddled with too many things to do?
- [ ] Need a friend to gently (and sometimes, not so gently) remind you of all your deadlines?

Then I've got the solution for **_you_**!

- a chatbot like no other
- sarcastic yet efficient 🤩
- scalding yet funny 🙄
- sciencey like no other!

With **MANY** functionalities, like:

1. Adding different types of tasks
2. Marking them complete (or incomplete)
3. Finding all tasks due on a certain date
4. And _more_...

And if you're a computer science enthusiast curious about the behind-the-scenes stuff, this is a glimpse of what the code looks like:

```
    public void start() {
        String userInput;
        ui.showWelcomeMessage();

        try {
            ArrayList<Task> loadedTasks = Storage.load(); 
            taskList.setTasks(loadedTasks); 
        } catch (IOException e) {
            System.out.println("Are you ready to put science first and put yourself to the test?");
        }

        while (true) {
            userInput = ui.readCommand();
            try {
                Commandable command = parser.parse(userInput, taskList, true); 
                command.execute();
                if (command instanceof ByeCommand) {
                    break; 
                }
            } catch (ViktorException e) {
                ui.showErrorMessage(e.getMessage());
            } catch (Exception e) {
                ui.showErrorMessage("Something went wrong! Very wrong! Hextech is alive!");
            }
        }

        ui.closeScanner();
    }

    public static void main(String[] args) {
        Viktor viktor = new Viktor();
        viktor.start();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Rae2480 committed:
Implement initial skeletal version of chatbot

Renamed chatbot from "Duke" to "Viktor".
Added a basic greeting as per requirements.

Rae2480 committed:
Implement echo and exit functions

Rae2480 committed:
Add customizations & credit AI

Both display and responses are now different from DUKE.
Responses are not randomized from a range of responses.
Observations wrt AI use recorded

Rae2480 committed:
Add storing in list & display function

Rae2480 committed:
Add list command

Rae2480 committed:
Add Task class; implement mark and unmark commands

Rae2480 committed:
Create child classes of Task; refactor Viktor class

Rae2480 committed:
Semi-automate testing; Handle errors

Rae2480 committed:
Add delete command

Rae2480 committed:
Replace if-else logics with enums

Rae2480 committed:
Updated AI.MD

Rae2480 committed:
Merge branch 'master' of https://github.com/Rae2480/ip

Rae2480 committed:
Implement saving and loading function

Rae2480 committed:
Edit testing scripts

Rae2480 committed:
Merge branch 'master' into branch-level-7

Rae2480 committed:
Add DateParser class and TIME command

Rae2480 committed:
Update README

Rae2480 committed:
Merge branch 'master' into branch-Level-8

Rae2480 committed:
Attempt to fix branching issues

Rae2480 committed:
Merge branch 'branch-Level-8'

Rae2480 committed:
Implement OOP paradigm

Rae2480 committed:
Update automated testing files and modify some responses

Rae2480 committed:
Organize into packages

Rae2480 committed:
Merge branch 'add-gradle-support'

Rae2480 committed:
Test all Task subclasses using JUnit and make tweaks

Rae2480 committed:
Create JAR file

Rae2480 committed:
Add Find functionality and corresponding class

Rae2480 committed:
Tweak code to comply with coding standards

Rae2480 committed:
Add JavaDoc comments

Rae2480 committed:
Update AI reflections

Rae2480 committed:
Merge branch 'branch-Level-9'

Rae2480 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/viktor/parser/Parser.java

Rae2480 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/viktor/commands/AddTaskCommand.java
#	src/main/java/viktor/tasks/Deadline.java
#	src/main/java/viktor/tasks/Event.java
#	src/main/java/viktor/tasks/Task.java
#	src/main/java/viktor/tasks/TaskType.java
#	src/main/java/viktor/tasks/Todo.java

Rae2480 committed:
Fix UI formatting

Rae2480 committed:
Change UI formatting

bipplane commented on src/main/java/viktor/commands/DeleteTaskComma.java:
```diff
@@ -7,16 +7,29 @@
+/**
+ * Command to delete a task from the task list.
+ */
 public class DeleteTaskComma implements Commandable {
```
The class name should be written in full for clarity, e.g., DeleteTaskCommand instead of DeleteTaskComma

bipplane commented on src/main/java/viktor/tasks/Event.java:
```diff
@@ -3,33 +3,64 @@
     private final LocalDateTime from;
     private final LocalDateTime to;
 
+    /**
+     * Constructs a Event task with a description, start and end date.
+     * 
+     * @param description The description of the task.
+     * @param from The start date of the task in string format.
+     * @param to The end date of the task in string format.
+     */
```
Great JavaDocs! Overall very clear and concise

bipplane commented on src/main/java/viktor/tasks/Deadline.java:
```diff
@@ -5,31 +5,61 @@
+/**
+ * Represents an deadline task with a description, and due time
+ */
 public class Deadline extends Task {
```
Same as above comment! Maybe something like DeadlineTask?

Rae2480 replied:
Ooh, may i ask why you think Deadline task is a better name? I'll definitely consider changing it tho, thanks! :D

bipplane replied:
It's mainly a naming convention, since being able to tell whether a class is extending another class purely based on the filename is a gamechanger when you have loads of files to look at

bipplane commented on src/main/java/viktor/commands/UnmarkCommand.java:
```diff
@@ -7,16 +7,29 @@
+     * @throws ViktorException If the input is invalid or there is an error with task creation.
+     */
     @Override
     public void execute() throws ViktorException {
```
This method should be including `IOException` in the exceptions thrown in this method! And remember to catch the thrown `VictorException` as well

Rae2480 replied:
Ooh thanks for the reminder! Appeciate it!

bipplane commented on src/main/java/viktor/commands/AddTaskCommand.java:
```diff
@@ -34,12 +35,30 @@ public class AddTaskCommand implements Commandable {
         this.isBeingTested = isBeingTested;
     }
 
+    /**
+     * Executes the command to add a task to the task list based on user input.
+     * 
+     * Creates a Todo, Deadline, or Event task and adds it to the task list, depending on user input.
+     * Prints a randomised response to the user, echoing the task description.
+     * During testing, responses are fixed.
+     * Saves saves the task list after the task is added.
+     * 
+     * @throws ViktorException If the input is invalid or there is an error with task creation.
+     */
```
Love the paragraphing here!

bipplane reviewed (commented):
Overall LGTM! Keep up the good work ❤️

Rae2480 committed:
Fix coding standard violations in JUnit tests

Rae2480 committed:
Create GUI

Rae2480 committed:
Add JUnit test for DateParser

Rae2480 committed:
Style GUI interface

Rae2480 committed:
Merge branch 'GUI'

curiousfun88 commented on src/main/java/viktor/commands/DeleteTaskComma.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Command to delete a task from the task list.
+ */
+public class DeleteTaskComma implements Commandable {
```
Since it is a class to house the delete task command, by DeleteTaskComma, do you mean DeleteTaskCommand instead?

curiousfun88 commented on src/main/java/viktor/parser/DateParser.java:
```diff
@@ -0,0 +1,35 @@
+    private static final DateTimeFormatter INPUT_DATE_FORMATTER = DateTimeFormatter.ofPattern("d/M/yyyy");
+    private static final DateTimeFormatter OUTPUT_DATE_FORMATTER = DateTimeFormatter.ofPattern("MMM d yyyy");
+    private static final DateTimeFormatter ALTERNATIVE_FORMATTER = DateTimeFormatter.ofPattern("MMM d yyyy, HH:mm");
+
```
LGTM. Just some minor recommendations: would it be better to standardize the number of lines left blank between every block? (For example, there are about 4 lines left blank here and there are no lines skipped between the LocalDateTime and formatDateTime methods)

Rae2480 replied:
Good catch! Will make the changes! Thanks for the thorough code review ☺️

curiousfun88 commented on src/main/java/viktor/ui/UI.java:
```diff
@@ -0,0 +1,38 @@
+
+public class UI {
+    public static final String LOGO = "\n" 
+            + "\t" + "\u001B[32m" + "      .__ __      __                \n"
```
The Viktor illustration looks aesthetically pleasing! Good job! LGTM!

curiousfun88 reviewed (commented):
Overall, your code is very complete and easy to read. It was a pleasure reviewing your code. You are also very in pace with the ip requirements. Good job and all the best!

curiousfun88 commented on src/main/java/viktor/commands/TimeCommand.java:
```diff
@@ -0,0 +1,66 @@
+            throw new ViktorException("Invalid date format! Use 'd/M/yyyy', e.g., '12/12/2025'.");
+        }
+
+        boolean found = false;
```
Additional Comment: would it better to name your boolean variable isFound, for example, to align with the coding standard for boolean variable names? :)

Rae2480 committed:
Fix checkstyle issues

Rae2480 committed:
Fix some formatting issues

Rae2480 committed:
Fix bugs in DeadlineTask and tweak UI elements

Rae2480 committed:
Add in-app help by loading sample data on first run

New users struggle to understand available commands without explicit guidance.

Without initial instructions, users might face a steep learning curve and be unsure how to interact with the app effectively.

Let's: Load sample data on the first run to display all available methods in a message.

Providing sample data ensures users can immediately see how the app functions, reducing confusion and improving onboarding experience.

The sample data is only loaded on the first run, preventing interference with existing user data while ensuring a smooth introduction to the app.

Rae2480 committed:
Improve typing syntax for user-friendliness

Currently, the command syntax strictly requires `/by` and `/from`, which can be unintuitive for users.

This needs to change because users may naturally type `by` or `from` without a preceding slash, leading to unnecessary errors and frustration.

Let's allow `by` and `from` without a slash in addition to `/by` and `/from`.

This is done to improve usability while maintaining backward compatibility with existing commands.

Now, users can input deadlines and events more naturally, reducing friction in task entry.

Rae2480 committed:
Add assertions to DateParser to document key assumptions

The DateParser class lacks explicit assertions to document critical assumptions about input validity and expected intermediate states.

Without assertions, incorrect or unexpected inputs may lead to silent failures or undefined behavior, making debugging more difficult.

Let's: add assert statements at key points in DateParser to ensure assumptions hold, such as valid date formats, non-null inputs, and expected parsing behaviour.

Using the built-in assert feature provides lightweight runtime checks that help catch logical errors early while maintaining performance in production (since assertions can be disabled).

Rae2480 committed:
Improve code quality and add comments for readability.

The current code lacks sufficient comments that explain what is going on for enhanced readability, and certain areas could be refactored for better structure and maintainability.

Improving code clarity helps future maintainers understand the logic more easily and reduces the risk of introducing bugs during modifications.

Let's: Refactor the code to enhance structure and maintainability while adding comments to explain key logic sections.

A well-structured codebase with clear comments ensures easier debugging, extension, and collaboration.

Rae2480 committed:
Replace old FindCommand logic with streams.

The current FindCommand implementation relies on a manual loop to filter tasks, making the code more verbose and less readable.

This needs to change to improve readability, maintainability, and efficiency by using a more functional approach.

Let's: replace the loop with Java Streams to filter matching tasks concisely.

Using streams allows for a declarative approach, improving clarity by directly expressing the filtering logic while reducing boilerplate code.

This refactor ensures that task searching remains efficient while enhancing the overall code quality.

Rae2480 committed:
Merge pull request #1 from Rae2480/branch-A-Assertions

Add assertions to DateParser to document key assumptions

Rae2480 committed:
Merge pull request #2 from Rae2480/branch-A-CodeQuality

Improve code quality and add comments for readability.

Rae2480 committed:
Merge pull request #3 from Rae2480/branch-A-Streams

Replace old FindCommand logic with streams.

Rae2480 committed:
Improve GUI to allow resizing

Rae2480 committed:
Add relative parsing for time inputs for event and deadline tasks

Rae2480 committed:
Rename task subclasses and improve error-handling

Rae2480 committed:
Update README to include user guide

Rae2480 committed:
Upload UI screenshot

Rae2480 committed:
Add new command: help for greater support

Rae2480 committed:
Refactor parser method to follow SLAP

Rae2480 committed:
Update README to include all commands

Rae2480 committed:
Rename UI.png to Ui.png

Rae2480 committed:
Update README.md

Rae2480 committed:
Fix consistency issues

Rae2480 committed:
Merge branch 'master' of https://github.com/Rae2480/ip

Rae2480 committed:
Improve UI by enabling better text wrapping

Rae2480 committed:
Create gradle.yml

Rae2480 committed:
Add style test files

Rae2480 committed:
Ensure correct permission for style check files

Rae2480 committed:
Fix style issues according to TP standards

Rae2480 committed:
Update gradle.yml to remove codecoverage checks

Rae2480 committed:
Update ReadMe
